src/k2/httpproxy/HTTPProxy.cpp

```diff
@@ -167,6 +167,34 @@ HTTPProxy::_handleTxnBegin(shd::TxnBeginRequest&& request){
         });
 }
 
+seastar::future<std::tuple<sh::Status, shd::WriteResponse>>
+HTTPProxy::_handleWrite(K2TxnHandle& txn, shd::WriteRequest&& request, dto::SKVRecord&& k2record) {
+    return txn.write(k2record, request.isDelete, static_cast<dto::ExistencePrecondition>(request.precondition))
+        .then([](WriteResult&& result) {
+            return MakeHTTPResponse<shd::WriteResponse>(sh::Status{.code = result.status.code, .message = result.status.message}, shd::WriteResponse{});
+        });
+}
+
+seastar::future<std::tuple<sh::Status, shd::WriteResponse>>
+HTTPProxy::_handlePartialUpdate(K2TxnHandle& txn, shd::WriteRequest&& request, dto::SKVRecord&& k2record, const std::shared_ptr<k2::dto::Schema>& k2Schema, const std::shared_ptr<shd::Schema>& shdSchema) {
+    dto::SKVRecord k2KeyRecord(request.collectionName, k2Schema);
+    shd::SKVRecord shdKeyRecord(request.collectionName, shdSchema, std::move(request.key), true);
+    try {
+        _shdRecToK2(shdKeyRecord, k2KeyRecord);
+    }  catch(shd::DeserializationError& err) {
+        return MakeHTTPResponse<shd::WriteResponse>(sh::Statuses::S400_Bad_Request(err.what()), shd::WriteResponse{});
+    }
+    std::vector<String> fieldsForPartialUpdate;
+    fieldsForPartialUpdate.reserve(request.fieldsForPartialUpdate.size());
+    for (sh::String& field: request.fieldsForPartialUpdate) {
+        fieldsForPartialUpdate.push_back(String(std::move(field)));
+    }
+    return txn.partialUpdate(k2record, fieldsForPartialUpdate, k2KeyRecord.getKey())
```

Comment:
We are trying to avoid using the interface with the key specified directly, so you should only pass a record and the fieldsforupdate. The record's key fields will be used to specify which key to update. This also avoids the problem you mentioned where the key fields can be changed.

Reply:
Changed key field and using key from record.